dot_config/doom/config.org
* ~config.el~

#+BEGIN_SRC :tangle "config.el"
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
#+END_SRC

This is where to place your private configuration.

#+BEGIN_QUOTE
You do not need to run ~doom sync~ after modifying this file.
#+END_QUOTE

Here are some additional functions/macros that could help you configure Doom:

- ~load~ for loading external *.el files relative to this one
- ~use-package~ for configuring packages
- ~after~ for running code after a package has loaded
- ~add-load-path~ for adding directories to the ~load-path~, relative to
  this file. Emacs searches the ~load-path~ when you load packages with
  ~require~ or ~use-package~.
- ~map~ for binding new keys

#+BEGIN_SRC :tangle no
To get information about any of these functions/macros, move the cursor over
the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
This will open documentation for it, including demos of how they are used.

You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
they are implemented.
#+END_SRC

** Personal Settings

Some functionality uses this to identify you, e.g. GPG configuration, email clients, file templates and snippets.

#+BEGIN_SRC emacs-lisp :tangle "config.el"
(setq user-full-name "Le Anh Duc"
      user-login-name "anhdle14"
      user-mail-address "anhdle14@icloud.com")

(setq auth-sources '("~/.authinfo.gpg")
      auth-source-cache-expiry nil)

(setq default-tab-width 2)
#+END_SRC

*** Default

#+BEGIN_SRC emacs-lisp :tangle "config.el"
(setq-default delete-by-moving-to-trash t) ; delete files to trash

(setq undo-limit 80000000            ; Raise undo-limit to 80Mb
      evil-want-fine-undo t          ; By default while in insert all changes are done in big blob. Be more granular
      auto-save-default t            ; Just auto-save because of auto-forget
      truncate-string-ellipsis "..." ; Unicode ellipsis are nicer than "...", and also save /precious/ space
      password-cache-expiry nil      ; I can trust my computers
      scroll-margin 2)
; Enable time in the mode-line
(display-time-mode 1)

; Iterate through CamelCase words
(global-subword-mode 1)
#+END_SRC

*** UI

There are two ways to load a theme. Both assume the theme is installed and available. You can either set ~doom-theme~ or manually load a theme with the ~load-theme~ function.

#+BEGIN_SRC emacs-lisp :tangle "config.el"
(setq doom-theme 'doom-gruvbox)
#+END_SRC

This determines the style of line numbers in effect. If set to `nil', line
numbers are disabled. For relative line numbers, set this to `relative'.

#+BEGIN_SRC emacs-lisp :tangle "config.el"
(setq display-line-numbers-type t)
#+END_SRC

Font is key!

#+BEGIN_SRC emacs-lisp :tangle "config.el"
(setq doom-font (font-spec :family "Iosevka" :size 14)
      doom-big-font (font-spec :family "Iosevka" :size 24)
      doom-variable-pitch-font (font-spec :family "Iosevka" :size 18)
      doom-unicode-font (font-spec :family "Iosevka")
      doom-serif-font (font-spec :family "Iosevka" :weight 'light))
#+END_SRC

** Org-mode

If you use `org' and don't want your org files in the default location below,
change `org-directory'. It must be set before org loads!

#+BEGIN_SRC emacs-lisp :tangle "config.el"
(setq org-directory "~/Developer/github.com/anhdle14/org")
#+END_SRC

* ~init.el~

This file controls what Doom modules are enabled and what order they load in. Remember to run 'doom sync' after modifying it!

#+BEGIN_SRC emacs-lisp :tangle "init.el"
;;; $DOOMDIR/init.el -*- lexical-binding: t; -*-
#+END_SRC

#+BEGIN_SRC emacs-lisp :tangle no
;; NOTE Press 'SPC h d h' (or 'C-h d h' for non-vim users) to access Doom's
;;      documentation. There you'll find a "Module Index" link where you'll find
;;      a comprehensive list of Doom's modules and what flags they support.
;;
;; NOTE Move your cursor over a module's name (or its flags) and press 'K' (or
;;      'C-c c k' for non-vim users) to view its documentation. This works on
;;      flags as well (those symbols that start with a plus).
;;
;;      Alternatively, press 'gd' (or 'C-c c d') on a module to browse its
;;      directory (for easy access to its source code).
#+END_SRC

Here doom-emacs defines all the packages per module that you can easily enable or disable via comment/uncomment or control the features of each packages via flags.

#+BEGIN_SRC emacs-lisp :tangle "init.el"
(doom! :input
       ;;chinese
       ;;japanese
       ;;layout            ; auie,ctsrnm is the superior home row

       :completion
       company           ; the ultimate code completion backend
       ;;helm              ; the *other* search engine for love and life
       ;;ido               ; the other *other* search engine...
       ;;ivy               ; a search engine for love and life
       vertico           ; the search engine of the future

       :ui
       ;;deft              ; notational velocity for Emacs
       doom              ; what makes DOOM look the way it does
       doom-dashboard    ; a nifty splash screen for Emacs
       doom-quit         ; DOOM quit-message prompts when you quit Emacs
       (emoji +unicode)  ; 🙂
       hl-todo           ; highlight TODO/FIXME/NOTE/DEPRECATED/HACK/REVIEW
       ;;hydra
       indent-guides     ; highlighted indent columns
       ligatures         ; ligatures and symbols to make your code pretty again
       ;;minimap           ; show a map of the code on the side
       modeline          ; snazzy, Atom-inspired modeline, plus API
       ;;nav-flash         ; blink cursor line after big motions
       ;;neotree           ; a project drawer, like NERDTree for vim
       ophints           ; highlight the region an operation acts on
       (popup +defaults)   ; tame sudden yet inevitable temporary windows
       tabs              ; a tab bar for Emacs
       treemacs          ; a project drawer, like neotree but cooler
       ;;unicode           ; extended unicode support for various languages
       vc-gutter         ; vcs diff in the fringe
       vi-tilde-fringe   ; fringe tildes to mark beyond EOB
       ;;window-select     ; visually switch windows
       workspaces        ; tab emulation, persistence & separate workspaces
       ;;zen               ; distraction-free coding or writing

       :editor
       (evil +everywhere); come to the dark side, we have cookies
       file-templates    ; auto-snippets for empty files
       fold              ; (nigh) universal code folding
       (format +onsave)  ; automated prettiness
       ;;god               ; run Emacs commands without modifier keys
       ;;lispy             ; vim for lisp, for people who don't like vim
       ;;multiple-cursors  ; editing in many places at once
       ;;objed             ; text object editing for the innocent
       ;;parinfer          ; turn lisp into python, sort of
       ;;rotate-text       ; cycle region at point between text candidates
       snippets          ; my elves. They type so I don't have to
       ;;word-wrap         ; soft wrapping with language-aware indent

       :emacs
       dired             ; making dired pretty [functional]
       electric          ; smarter, keyword-based electric-indent
       ;;ibuffer         ; interactive buffer management
       undo              ; persistent, smarter undo for your inevitable mistakes
       vc                ; version-control and Emacs, sitting in a tree

       :term
       ;;eshell            ; the elisp shell that works everywhere
       ;;shell             ; simple shell REPL for Emacs
       ;;term              ; basic terminal emulator for Emacs
       vterm             ; the best terminal emulation in Emacs

       :checkers
       syntax              ; tasing you for every semicolon you forget
       ;;(spell +flyspell) ; tasing you for misspelling mispelling
       ;;grammar           ; tasing grammar mistake every you make

       :tools
       ;;ansible
       ;;biblio            ; Writes a PhD for you (citation needed)
       debugger          ; FIXME stepping through code, to help you add bugs
       direnv
       docker
       editorconfig      ; let someone else argue about tabs vs spaces
       ein               ; tame Jupyter notebooks with emacs
       (eval +overlay)     ; run code, run (also, repls)
       ;;gist              ; interacting with github gists
       lookup              ; navigate your code and its documentation
       lsp               ; M-x vscode
       magit             ; a git porcelain for Emacs
       make              ; run make tasks from Emacs
       ;;pass              ; password manager for nerds
       pdf               ; pdf enhancements
       ;;prodigy           ; FIXME managing external services & code builders
       rgb               ; creating color strings
       taskrunner        ; taskrunner for all your projects
       terraform         ; infrastructure as code
       tmux              ; an API for interacting with tmux
       ;;upload            ; map local to remote projects via ssh/ftp

       :os
       (:if IS-MAC macos)  ; improve compatibility with macOS
       (tty +osc)          ; improve the terminal Emacs experience

       :lang
       ;;agda              ; types of types of types of types...
       ;;beancount         ; mind the GAAP
       ;;cc                ; C > C++ == 1
       ;;clojure           ; java with a lisp
       common-lisp       ; if you've seen one lisp, you've seen them all
       ;;coq               ; proofs-as-programs
       ;;crystal           ; ruby at the speed of c
       ;;csharp            ; unity, .NET, and mono shenanigans
       data              ; config/data formats
       (dart +flutter)   ; paint ui and not much else
       ;;dhall
       ;;elixir            ; erlang done right
       ;;elm               ; care for a cup of TEA?
       emacs-lisp        ; drown in parentheses
       ;;erlang            ; an elegant language for a more civilized age
       ;;ess               ; emacs speaks statistics
       ;;factor
       ;;faust             ; dsp, but you get to keep your soul
       ;;fortran           ; in FORTRAN, GOD is REAL (unless declared INTEGER)
       ;;fsharp            ; ML stands for Microsoft's Language
       ;;fstar             ; (dependent) types and (monadic) effects and Z3
       ;;gdscript          ; the language you waited for
       (go +lsp)         ; the hipster dialect
       ;;(haskell +lsp)    ; a language that's lazier than I am
       ;;hy                ; readability of scheme w/ speed of python
       ;;idris             ; a language you can depend on
       json              ; At least it ain't XML
       (java +meghanada) ; the poster child for carpal tunnel syndrome
       javascript        ; all(hope(abandon(ye(who(enter(here))))))
       julia             ; a better, faster MATLAB
       ;;kotlin            ; a better, slicker Java(Script)
       latex             ; writing papers in Emacs has never been so fun
       ;;lean              ; for folks with too much to prove
       ;;ledger            ; be audit you can be
       ;;lua               ; one-based indices? one-based indices
       markdown          ; writing docs for people to ignore
       ;;nim               ; python + lisp at the speed of c
       ;;nix               ; I hereby declare "nix geht mehr!"
       ocaml             ; an objective camel
       (org +pretty
            +dragndrop
            +gnuplot
            +hugo
            +journal
            +jupyter
            +noter
            +pandoc
            +roam2)      ; organize your plain life in plain text
       ;;php               ; perl's insecure younger brother
       plantuml          ; diagrams for confusing people more
       ;;purescript        ; javascript, but functional
       (python +lsp
               +pyright
               +pyenv
               +cpython) ; beautiful is better than ugly
       ;;qt                ; the 'cutest' gui framework ever
       ;;racket            ; a DSL for DSLs
       ;;raku              ; the artist formerly known as perl6
       rest              ; Emacs as a REST client
       ;;rst               ; ReST in peace
       ;;(ruby +rails)     ; 1.step {|i| p "Ruby is #{i.even? ? 'love' : 'life'}"}
       rust              ; Fe2O3.unwrap().unwrap().unwrap().unwrap()
       ;;scala             ; java, but good
       ;;(scheme +guile)   ; a fully conniving family of lisps
       sh                ; she sells {ba,z,fi}sh shells on the C xor
       ;;sml
       ;;solidity          ; do you need a blockchain? No.
       ;;swift             ; who asked for emoji variables?
       ;;terra             ; Earth and Moon in alignment for performance.
       web               ; the tubes
       yaml              ; JSON, but readable
       ;;zig               ; C, but simpler

       :email
       ;;(mu4e +org +gmail)
       ;;notmuch
       ;;(wanderlust +gmail)

       :app
       ;;calendar
       ;;emms
       ;;everywhere        ; *leave* Emacs!? You must be joking
       ;;irc               ; how neckbeards socialize
       ;;(rss +org)        ; emacs as an RSS reader
       ;;twitter           ; twitter client https://twitter.com/vnought

       :config
       literate
       (default +bindings +smartparens))
#+END_SRC

#+RESULTS:

* ~packages.el~

#+BEGIN_SRC :tangle "packages.el"
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
#+END_SRC

To install a package with Doom you must declare them here and run ~doom sync~ on the command line, then restart Emacs for the changes to take effect -- or use ~M-x doom/reload~.

To install SOME-PACKAGE from MELPA, ELPA, or emacsmirror:
#+BEGIN_SRC emacs-lisp :tangle no
(package! some-package)
#+END_SRC

To install a package directly from a remote git repo, you must specify a ~:recipe~. You will find documentation on what ~:recipe~ accepts here:

#+BEGIN_SRC emacs-lisp :tangle no
(package! another-package
  :recipe (:host github :repo "username/repo"))
#+END_SRC

If the package you are trying to install does not contain a ~PACKAGENAME.el~ file, or is located in a subdirectory of the repo, you will need to specify ~:files~ in the ~:recipe~:

#+BEGIN_SRC emacs-lisp :tangle no
(package! this-package
  :recipe (:host github :repo "username/repo"
           :files ("some-file.el" "src/lisp/*.el")))
#+END_SRC

If you would like to disable a package included with Doom, you can do so here with the ~:disable~ property:

#+BEGIN_SRC emacs-lisp :tangle no
(package! builtin-package :disable t)
#+END_SRC

You can override the recipe of a built in package without having to specify all the properties for ~:recipe~. These will inherit the rest of its recipe from Doom or MELPA/ELPA/Emacsmirror:

#+BEGIN_SRC emacs-lisp :tangle no
(package! builtin-package :recipe (:nonrecursive t))
(package! builtin-package-2 :recipe (:repo "myfork/package"))
#+END_SRC

Specify a ~:branch~ to install a package from a particular branch or tag. This is required for some packaages whose default branch isn't ~master~ which our package manager can't deal with.

#+BEGIN_SRC emacs-lisp :tangle no
(package! builtin-package :recipe (:branch "develop"))
#+END_SRC

Use ~:pin~ to specify a particular commit to install.

#+BEGIN_SRC emacs-lisp :tangle no
(package! builtin-package :pin "1a2b3c4d5e")
#+END_SRC

Doom's packages are pinned to a specific commit and updated from release to release. The ~unpin!~ macro allows you to unpin single packages...

#+BEGIN_SRC emacs-lisp :tangle no
(unpin! package)
;; or multiple packages
(unpin! package-1 package-2)
;; or all
(unpin! t)
#+END_SRC

** Emacs Application Framework

This doesn't work for now

#+begin_src emacs-lisp :tangle no
(when (package! eaf :recipe (:host github
                             :repo "emacs-eaf/emacs-application-framework"
                             :files ("*.el" "*.py" "app" "core")
                             :build (:not compile)))

  (package! ctable :recipe (:host github :repo "kiwanami/emacs-ctable"))
  (package! deferred :recipe (:host github :repo "kiwanami/emacs-deferred"))
  (package! epc :recipe (:host github :repo "kiwanami/emacs-epc")))

(use-package! eaf
  :commands (eaf-open-browser eaf-open find-file)
  :config
  (use-package! ctable)
  (use-package! deferred)
  (use-package! epc))
#+end_src


dot_config/doom/config.el
(setq user-full-name "Le Anh Duc"
      user-login-name "anhdle14"
      user-mail-address "anhdle14@icloud.com")

(setq auth-sources '("~/.authinfo.gpg")
      auth-source-cache-expiry nil)

(setq default-tab-width 2)

(setq-default delete-by-moving-to-trash t) ; delete files to trash

(setq undo-limit 80000000            ; Raise undo-limit to 80Mb
      evil-want-fine-undo t          ; By default while in insert all changes are done in big blob. Be more granular
      auto-save-default t            ; Just auto-save because of auto-forget
      truncate-string-ellipsis "..." ; Unicode ellipsis are nicer than "...", and also save /precious/ space
      password-cache-expiry nil      ; I can trust my computers
      scroll-margin 2)
; Enable time in the mode-line
(display-time-mode 1)

; Iterate through CamelCase words
(global-subword-mode 1)

(setq doom-theme 'doom-gruvbox)

(setq display-line-numbers-type t)

(setq doom-font (font-spec :family "Iosevka" :size 14)
      doom-big-font (font-spec :family "Iosevka" :size 24)
      doom-variable-pitch-font (font-spec :family "Iosevka" :size 18)
      doom-unicode-font (font-spec :family "Iosevka")
      doom-serif-font (font-spec :family "Iosevka" :weight 'light))

(setq org-directory "~/Developer/github.com/anhdle14/org")
